Add Envoy TCP Routing example demonstrating source-based dynamic cluster selection

SiiiTschiii opened the pull request:
Introduce a new Proxy-Wasm TCP filter example that demonstrates dynamic routing of TCP connections based on source IP addresses at the TCP stream context (L4). This example showcases application-layer routing decisions without HTTP processing, filling a gap in the existing HTTP-focused examples.

Inspired by the [wasmerang](https://github.com/SiiiTschiii/wasmerang) project, this example demonstrates advanced TCP routing patterns in Envoy/Istio/K8s environments using WASM filters.

## Key Features

- **Source-based routing**: Dynamic cluster selection based on source IP last octet (even → egress-router1, odd → egress-router2)
- **TCP stream context**: Operates at L4 rather than HTTP layer for protocol-agnostic routing
- **Envoy filter state API**: Uses `set_envoy_filter_state` foreign function with protobuf encoding
- **Complete Docker setup**: Includes Docker Compose environment for testing with httpbin.org
- **Comprehensive documentation**: Usage examples, expected output, and technical context

## Implementation Details

- Protobuf definitions for Envoy's filter state API with build-time code generation
- MSRV-compatible dependencies (Rust 1.65) with explicit pinning for CI compatibility
- Envoy v1.34.1 configuration with dual upstream clusters

## Documentation Highlights

- Links to Proxy-Wasm spec for TCP stream context
- Explanation of TCP vs HTTP layer routing trade-offs
- Future enhancement ideas: PROXY protocol support for source-based egress routing in managed K8s environments

This example is useful for scenarios requiring performance-critical or protocol-agnostic routing decisions where application-layer processing should be avoided.

SiiiTschiii committed:
Add TCP rerouting example with dynamic cluster selection

Introduce a new Proxy-Wasm TCP filter example that demonstrates dynamic
routing of TCP connections based on source IP addresses. This example
showcases TCP-level filtering, filling a gap in the existing HTTP-focused
examples.

Key features:
- Dynamic cluster selection based on source IP last octet (even/odd)
- Uses Envoy's set_envoy_filter_state foreign function with protobuf encoding
- Includes comprehensive Docker Compose setup for testing
- Full documentation with usage examples and expected output

Implementation details:
- Protobuf definitions for Envoy's filter state API
- Build script for code generation from proto files
- Unit tests for IP parsing and routing logic
- Envoy configuration with dual upstream clusters

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Fix formatting and clippy warning in tcp_rerouting example

- Remove trailing whitespace
- Use next_back() instead of last() for better performance

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Update Bazel dependencies

- Update log 0.4.27 → 0.4.28
- Update proc-macro2 1.0.101 → 1.0.103
- Update quote 1.0.41 → 1.0.42
- Update unicode-ident 1.0.19 → 1.0.22

Generated by: bazelisk run //bazel/cargo:crates_vendor -- --repin all

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Add TCP rerouting example to main README

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

mmorel-35 committed:
Update MSRV to Rust v1.68.0. (#335)

This version is required by the quote crate (a dependency of mockalloc).

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Update bazel-skylib to v1.8.2. (#336)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

SiiiTschiii committed:
Add Apache 2.0 license headers to tcp_rerouting example files

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Pin tcp_rerouting build dependencies for Rust 1.65 MSRV compatibility

Pin proc-macro2 and quote in build-dependencies to versions compatible
with Rust 1.65 (the SDK's MSRV). Without these pins, cargo may select
newer versions (e.g., quote 1.0.42) that require Rust 1.68+, causing
CI builds to fail.

This follows best practices by explicitly pinning problematic transitive
dependencies rather than setting rust-version, which is not used by
other examples.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

PiotrSikora commented on examples/tcp_rerouting/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+description = "Proxy-Wasm plugin example: TCP Rerouting based on source IP"
+license = "Apache-2.0"
+edition = "2021"
+rust-version = "1.65"
```
Does this work? We had to bump MSRV in the SDK because `cargo` always selected the latest version of `quote` crate, ignoring SDK's `rust-version`.

PiotrSikora commented on examples/tcp_rerouting/README.md:
```diff
@@ -0,0 +1,81 @@
+
+Proxy-Wasm TCP filter that dynamically routes connections to different upstream clusters based on the source IP address.
+
+Most WASM filter examples focus on HTTP, but this example shows how to work at the TCP/IP level.
```
While this statement is true, it probably doesn't need to be permanently inscribed in the README.

SiiiTschiii replied:
Adressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332/commits/e68a1655a068cc4742c80ccafa453655fd50125c and https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332/commits/846ad4bd59d4a53c6487befefc95a6d981ec9da3

PiotrSikora reviewed (changes requested):
Thanks, this is great real-world example!

SiiiTschiii committed:
Add license header to .proto file and refine MSRV approach

- Add Apache 2.0 license header to set_envoy_filter_state.proto
- Remove rust-version field from tcp_rerouting example Cargo.toml
- Keep explicit dependency pins for proc-macro2 and quote to ensure
  Rust 1.65 compatibility

This approach follows the pattern of other examples (no rust-version)
while ensuring MSRV compatibility through explicit dependency pinning
rather than relying on rust-version, which doesn't prevent Cargo from
selecting incompatible dependency versions.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Remove generated protobuf file from version control

Remove examples/tcp_rerouting/src/generated/ from git tracking and
add to .gitignore. Generated code should not be committed as it:
- Can be regenerated from the .proto file via build.rs
- Creates unnecessary diff noise when dependencies update
- Can get out of sync with source

The file is automatically generated during build by prost-build,
so this does not affect functionality.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Rename example to "Envoy TCP Routing"

Rename tcp_rerouting to envoy_tcp_routing to:
- Clearly indicate this is an Envoy-specific example
- Use "routing" instead of "rerouting" for better terminology

Changes:
- Renamed directory: examples/tcp_rerouting → examples/envoy_tcp_routing
- Updated package name: proxy-wasm-example-envoy-tcp-routing
- Updated all references in README.md files
- Updated Docker Compose network name and WASM filename
- Updated Envoy config filter names and WASM path
- Updated .gitignore path

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Remove unnecessary statement from README

Remove the statement about most examples focusing on HTTP, as it's
not essential information for the README.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Clarify TCP stream context operation in README

Add explanation that the example operates at the TCP stream context
(L4) rather than the HTTP application layer, highlighting its
usefulness for scenarios where application-layer processing should
be avoided.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Add Proxy-Wasm spec links and mention performance benefits

Add links to TCP stream and HTTP stream contexts in the Proxy-Wasm
spec, and clarify that operating at the TCP level is useful for
performance and protocol-agnostic routing decisions.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
docs(envoy_tcp_routing): improve README with destination info and future enhancements

- Fix grammar: "In separate terminals" → "In a separate terminal"
- Add consistency: "routes to" → "routes via" in output examples
- Add "Note on Destination Information" section explaining:
  - Hardcoded httpbin.org endpoints in both clusters
  - PROXY protocol for preserving metadata across TCP hops
  - Implementation approach using WASM stream context + Envoy config
- Document future enhancement for source-based egress router use case:
  - Routing via different external IPs (e.g., even vs. odd source IP)
  - Explain conventional network approach (routing tables + IP rules)
  - Note challenges in managed K8s (requires plugins like Multus CNI)
  - Clarify how K8s abstracts layer 3 routing configuration

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,81 @@
+
+### Test the Routing
+
+In separate terminals, test the routing behavior with different source IP addresses:
```
This was related to the previous step where you do `docker compose up` in non-detached mode. So at least another terminal would be needed when doing this step. Updated it to make it more clear. https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332/commits/d66580052b06cf51cdcdd3294e23b4b08dd5f78a

SiiiTschiii committed:
Revert "Update Bazel dependencies"

This reverts commit 467e8d60249357747a2c22de511e77b2c4f40f36.

The Bazel dependency updates are not related to the envoy_tcp_routing
example, which builds using cargo with wasm32-wasip1 target. The example
has its own dependency pins in Cargo.toml for MSRV compatibility and does
not use the Bazel build system.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii commented:
@PiotrSikora thanks for the review. I believe I addressed all of them.

SiiiTschiii committed:
Switch TCP routing example to wasm32-wasi target and pin MSRV-compatible deps

- Change build target from wasm32-wasip1 to wasm32-wasi for MSRV (1.65) compatibility
- Update docker-compose.yaml to mount wasm32-wasi artifact path
- Pin build dependencies to latest MSRV-compatible versions:
  - prost-build 0.11.9, quote 1.0.41, proc-macro2 1.0 (existing pins)
  - indexmap 2.11.4, home 0.5.5 (pinned transitive deps)
- Update README.md build instructions to reflect new target

The wasm32-wasip1 target is not available in Rust 1.65. By switching to
wasm32-wasi and constraining transitive dependencies that require newer
Rust editions (indexmap 2.12+ needs edition 2024, home 0.5.12+ needs
edition 2024), we ensure the example builds with the project's MSRV.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

mmorel-35 committed:
Update rules_cc to v0.2.14. (#337)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

SiiiTschiii commented:
@PiotrSikora Note on cargo outdated CI Check that is failing

The cargo outdated check in CI currently fails for this example because it runs with the latest stable Rust (1.83) rather than the project's MSRV (1.65). When analyzing available updates, it encounters transitive dependencies like home 0.5.12 which require edition 2024 (not yet stabilized in Rust 1.83).

The example itself builds successfully with MSRV 1.65 because we've pinned the problematic transitive dependencies (indexmap = "=2.11.4", home = "=0.5.5"). However, cargo outdated doesn't respect these pins when checking the crates.io index for newer versions—it attempts to parse manifests of all available versions and fails on edition 2024 crates.

Potential solutions:

1. Run the cargo outdated CI step with the MSRV toolchain (1.65)
2. Skip the outdated check for examples that explicitly manage MSRV-constrained dependencies
3. Document this as a known limitation when maintaining MSRV-compatible examples

The dependency pins are intentional and necessary to maintain MSRV compatibility while still using the latest compatible versions of direct dependencies.

mmorel-35 committed:
Update rules_rust to v0.67.0 (with Rust v1.90.0). (#334)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Update Bazel to v7.7.1. (#338)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Add support for Bazel modules. (#333)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

PiotrSikora committed:
Update cargo-audit to v0.22.0. (#339)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update cargo-outdated to v0.17.0. (#340)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

SiiiTschiii committed:
Update TCP routing example to use latest dependencies

- Upgrade prost to 0.14 (from 0.11.9)
- Remove MSRV-specific dependency pins (indexmap, home, quote)
- Switch back to wasm32-wasip1 target (now supported in current Rust)
- Update docker-compose.yaml to use wasm32-wasip1 artifact path

This change addresses the cargo outdated CI check failures and aligns
with the maintainer's guidance that examples don't need to be MSRV-compatible.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Update TCP routing example to use latest dependencies

- Upgrade prost to 0.14 (from 0.11.9)
- Remove MSRV-specific dependency pins (indexmap, home, quote)
- Switch back to wasm32-wasip1 target (now supported in current Rust)
- Update docker-compose.yaml to use wasm32-wasip1 artifact path

This change addresses the cargo outdated CI check failures and aligns
with the maintainer's guidance that examples don't need to be MSRV-compatible.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

mmorel-35 committed:
Update rules_rust to v0.68.1 (with Rust v1.91.1). (#343)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Add support for reproducible metadata in Bazel modules. (#344)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Add Dependabot. (#342)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

dependabot[bot] committed:
Update actions/checkout to v6. (#346)

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 6.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v6)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Update actions/cache to v5. (#347)

Bumps [actions/cache](https://github.com/actions/cache) from 4 to 5.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Update actions/setup-go to v6. (#348)

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 5 to 6.
- [Release notes](https://github.com/actions/setup-go/releases)
- [Commits](https://github.com/actions/setup-go/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/setup-go
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

PiotrSikora committed:
Update addlicense to v1.2.0. (#349)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update buildifier to latest (2026-01-21). (#350)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update Bazelisk to v1.28.1. (#351)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update Bazel to v8.5.1. (#352)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

SiiiTschiii committed:
Update Cargo.toml copyright header format

Change copyright header from C++ style (//) comments to TOML style (#)
comments and update year to 2026.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Update Cargo.toml copyright header format

Change copyright header from C++ style (//) comments to TOML style (#)
comments and update year to 2026.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

PiotrSikora commented:
> @PiotrSikora alright I added the missing license. Can I run those blocked CI steps locally myself?

Yes, see: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/main/DEVELOPMENT.md#testing

Although, some Google org-wide checks like: CLA, DCO and Actions Workflow Security Scan aren't there.

SiiiTschiii commented:
@PiotrSikora ok now it looks good. Let's get that merged if you are still ok with it.

mmorel-35 committed:
Add workflow for publishing releases to Bazel Central Registry. (#341)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

mmorel-35 committed:
Use "proxy-wasm-rust-sdk" as the Bazel module name. (#355)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

PiotrSikora committed:
Fix instructions for updating Bazel dependencies. (#356)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update Go to v1.24. (#357)

This is the minimum version required by Bazelisk v1.28.1.

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

dependabot[bot] committed:
Update getrandom to v0.4 in examples. (#353)

Updates the requirements on [getrandom](https://github.com/rust-random/getrandom) to permit the latest version.
- [Changelog](https://github.com/rust-random/getrandom/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rust-random/getrandom/compare/v0.3.0...v0.4.1)

---
updated-dependencies:
- dependency-name: getrandom
  dependency-version: 0.4.1
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

PiotrSikora committed:
Release v0.2.4 for Bazel Central Registry. (#354)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Update version to v0.2.5-dev. (#359)

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Make "mockalloc" an optional feature. (#358)

Those tests interfere with regular tests and benchmarks,
because they install a custom global allocator.

Furthermore, this allows us to keep lower MSRV.

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora commented:
@SiiiTschiii the "whole file rewrite" of `.github/workflows/rust.yml` makes it pretty unreadable for a PR, and apparently broke other examples, but it's really hard to tell what's the cause from looking at the workflow file.

SiiiTschiii committed:
Add envoy_tcp_routing example to CI workflow

- Add envoy_tcp_routing to examples and reactors matrix
- Install protobuf compiler for envoy_tcp_routing builds
- Clean generated protobuf files before build to avoid cache issues
- Update Envoy validation to mount example-specific WASM files
- Move envoy.yaml to example root directory
- Add .gitignore for generated protobuf code

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Add envoy_tcp_routing example to CI workflow

- Add envoy_tcp_routing to examples and reactors matrix
- Install protobuf compiler for envoy_tcp_routing builds
- Clean generated protobuf files before build to avoid cache issues
- Update Envoy validation to mount example-specific WASM files
- Move envoy.yaml to example root directory
- Add .gitignore for generated protobuf code

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii commented:
> @SiiiTschiii the "whole file rewrite" of `.github/workflows/rust.yml` makes it pretty unreadable for a PR, and apparently broke other examples, but it's really hard to tell what's the cause from looking at the workflow file.

@PiotrSikora My bad. I didn't check the diff. I removed the formatting changes and also fixed a bug that broke the other examples. Thanks for running the pipeline again!

SiiiTschiii committed:
Fix missing protobuf compiler steps in reactors CI job

The reactors job was missing the 'Install protobuf compiler' and
'Clean generated files' steps that were only added to the examples
job, causing the build to fail for envoy_tcp_routing.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Fix missing protobuf compiler steps in reactors CI job

The reactors job was missing the 'Install protobuf compiler' and
'Clean generated files' steps that were only added to the examples
job, causing the build to fail for envoy_tcp_routing.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii commented:
@PiotrSikora and another try.

SiiiTschiii committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

mmorel-35 committed:
Fix presubmit.yml for Bazel Central Registry. (#360)

Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

PiotrSikora committed:
Use rust_shared_library for building Wasm modules with Bazel. (#362)

The rust_binary(out_binary=True) hack is no longer needed,
so we can remove the patch for it.

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

SiiiTschiii committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

SiiiTschiii commented:
@PiotrSikora can you please trigger the pipeline again. Thank you.

SiiiTschiii commented:
@PiotrSikora can I ask you to trigger the pipeline again. Thank you

SiiiTschiii commented:
@PiotrSikora is there anything else I need to change? Pipeline looks green. Otherwise I'd appreciate to get this one merged. 🙌

SiiiTschiii commented:
@PiotrSikora can we merge this PR please? Or is there anything left to do? Thank you!

PiotrSikora commented on .github/workflows/rust.yml:
```diff
@@ -455,6 +456,18 @@ jobs:
+      if: matrix.example == 'envoy_tcp_routing'
+      run: |
+        rm -rf src/generated
+        rm -rf target
```
I don't think we need to cleanup `target` directory, since it's just cache.

SiiiTschiii replied:
Adressed.

PiotrSikora commented on .github/workflows/rust.yml:
```diff
@@ -529,6 +543,18 @@ jobs:
+      if: matrix.example == 'envoy_tcp_routing'
+      run: |
+        rm -rf src/generated
+        rm -rf target
```
Same here.

SiiiTschiii replied:
Addressed

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+
+    // Parse as u8
+    last_segment.parse::<u8>().ok()
+}
```
This parsing won't work for IPv6 addresses, why not parse using `std::net::SocketAddr` and `addr.parse()`?

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+
+impl StreamContext for TcpReroutingStream {
+    fn on_new_connection(&mut self) -> Action {
+        if let Some(source_addr_bytes) = self.get_property(vec!["source", "address"]) {
```
This could more more readable with "let else" flow instead of deeply nested `if` statements, i.e.:
```rust
let Some(source_addr_bytes) = self.get_property(vec!["source", "address"]) else {
    return Action::Continue;
}

let Ok(source_addr) = std::str::from_utf8(&source_addr_bytes) else {
    return Action::Continue;
}

info!("[TCP WASM] Source address: {}", source_addr);

let Some(last_octet) = extract_last_octet(source_addr) else {
    return Action::Continue;
}

info!("[TCP WASM] Source IP last octet: {}", last_octet);

...
```

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+                // Extract the last octet from the source IP address
+                if let Some(last_octet) = extract_last_octet(source_addr) {
+                    info!(
+                        "[TCP WASM] Source IP last octet: {}, intercepting ALL traffic",
```
Please drop "intercepting ALL traffic", since it's misleading and nothing is in fact being intercepted.

Also, the traffic is going to be routed to `egress-router1` (the default cluster) even if we don't reach this stage.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+                    info!(
+                        "[TCP WASM] set_envoy_filter_state status (envoy.tcp_proxy.cluster): {:?}",
+                        status
+                    );
```
Perhaps turn this into `warn` on failure? I don't think we want to log success here.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+                    info!("[TCP WASM] Routing to {}", cluster);
+
+                    // Set the cluster via Envoy's filter state mechanism using proper protobuf encoding
+                    use set_envoy_filter_state::{LifeSpan, SetEnvoyFilterStateArguments};
```
Nit: move this to global imports.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+                        "[TCP WASM] set_envoy_filter_state status (envoy.tcp_proxy.cluster): {:?}",
+                        status
+                    );
+                    info!("[TCP WASM] Rerouting to {} via filter state", cluster);
```
We already log `[TCP WASM] Routing to` upon cluster selection in line 89, so this seems redundant. Perhaps leave only this line?

Also, other than this final line (which could include source address, e.g. `"Routing {source_address} to {cluster}"), other logs should be probably done at `debug` or even `trace` levels.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/9273dcdb0212abeb88586d75aa9f52f58abec832

PiotrSikora commented on examples/envoy_tcp_routing/build.rs:
```diff
@@ -0,0 +1,25 @@
+        .out_dir(out_dir)
+        .compile_protos(&["src/set_envoy_filter_state.proto"], &["src/"])
+        .unwrap();
+    println!("cargo:rerun-if-changed=src/set_envoy_filter_state.proto");
```
Nit: `cargo:rerun-if-changed` lines are usually logged at the beginning of `main()`.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/0a6e688e768c8ae8f7511a0597e16bd958005816

PiotrSikora commented on examples/envoy_tcp_routing/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+[build-dependencies]
+prost-build = "0.14"
+proc-macro2 = "1.0"
+
```
Nit: empty line at the end.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/0a6e688e768c8ae8f7511a0597e16bd958005816

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+    image: envoyproxy/envoy:v1.34.1
+    entrypoint: /usr/local/bin/envoy -c /etc/envoy.yaml -l info --service-cluster proxy
+    volumes:
+      - ./envoy.yaml:/etc/envoy.yaml
```
This should match other examples, i.e. `./envoy.yaml:/etc/envoy/envoy.yaml`

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+
+services:
+  proxy:
+    image: envoyproxy/envoy:v1.34.1
```
Please use `envoyproxy/envoy:v1.34-latest` to match other examples

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+services:
+  proxy:
+    image: envoyproxy/envoy:v1.34.1
+    entrypoint: /usr/local/bin/envoy -c /etc/envoy.yaml -l info --service-cluster proxy
```
You could use this instead to avoid overriding the path to the entrypoint binary:
```
command:
- "--config-path /etc/envoy/envoy.yaml"
- "--log-level info"
- "--service-cluster proxy"
```
(although, I think this matches the default anyway?)

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/envoy.yaml:
```diff
@@ -0,0 +1,90 @@
+          port_value: 10000
+      filter_chains:
+        - filters:
+            # WASM filter for TCP rerouting
```
Nit: `rerouting` -> `routing`.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,89 @@
+
+This example demonstrates how to build a TCP filter that:
+
+- Intercepts incoming TCP connections
```
Nit: `intercepts` -> `inspects`

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/0a6e688e768c8ae8f7511a0597e16bd958005816

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+      - "8001:8001"
+
+networks:
+  envoymesh:
```
Perhaps rename this network to avoid collisions with other examples that don't create specific subnet?

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/envoy.yaml:
```diff
@@ -0,0 +1,90 @@
+  address:
+    socket_address:
+      address: 0.0.0.0
+      port_value: 8001
```
This isn't used anywhere.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+      - envoymesh
+    ports:
+      - "10000:10000"
+      - "8001:8001"
```
We don't use/expose admin port.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+    ports:
+      - "10000:10000"
+      - "8001:8001"
+
```
Please add local `httpbin` backend:
```
  httpbin:
    image: mccutchen/go-httpbin
    hostname: httpbin
    ports:
      - "8080:8080"
    networks:
      - envoymesh
```
and add dependency on it to the `envoy` service.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
````diff
@@ -0,0 +1,89 @@
+
+```console
+$ docker compose logs -f
+```
````
Nit: Please use "Expected Envoy logs:" to match other examples.

We don't deamonize `docker compose up`, so the logs are already logged in the foreground, and we don't need `docker compose logs`.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/0a6e688e768c8ae8f7511a0597e16bd958005816

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,32 @@
+# limitations under the License.
+
+services:
+  proxy:
```
Nit: `proxy` -> `envoy` to match other examples.

SiiiTschiii replied:
Addressed in https://github.com/proxy-wasm/proxy-wasm-rust-sdk/commit/b3a3156632a1477bd03d1fb9cd9c6b36d5bd7068

SiiiTschiii committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

SiiiTschiii committed:
ci: stop cleaning target in envoy_tcp_routing jobs

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232069095

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232069785
Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
example(envoy_tcp_routing): align compose and envoy config with repo conventions

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232289396

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232301609

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232303872

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232607674

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232547085

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232547725

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232531138

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232307598

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232545105
Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
example(envoy_tcp_routing): refactor tcp routing flow and logging

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232096102

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232140544

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232155132

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232169015

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232172142

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232219799
Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
docs(example/envoy_tcp_routing): align README and build metadata with review nits

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232316706

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232564405

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232226333

Addresses: https://github.com/proxy-wasm/proxy-wasm-rust-sdk/pull/332#discussion_r3232231562
Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
ci: restore rust workflow formatting

Keep only the intended removal of target cleanup in envoy_tcp_routing jobs.

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Fix network configuration for envoy-tcp-routing example

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Update envoy.yaml configuration

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
examples/envoy_tcp_routing: use latest go-httpbin image

Signed-off-by: Christof Gerber <christof.gerber1@gmail.com>

SiiiTschiii committed:
Merge branch 'main' into SiiiTschiii/example-tcp-re-routing

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,166 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! TCP Rerouting Example
```
Style: `TCP Rerouting example` -> `TCP routing example`

PiotrSikora commented on examples/envoy_tcp_routing/src/lib.rs:
```diff
@@ -0,0 +1,166 @@
+    });
+}}
+
+struct TcpReroutingRoot;
```
Style: `TcpReroutingRoot` -> `TcpRoutingRoot` (here and throughout the example)

PiotrSikora commented on examples/envoy_tcp_routing/envoy.yaml:
```diff
@@ -0,0 +1,83 @@
+              typed_config:
+                "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
+                stat_prefix: destination
+                cluster: egress-router1 # Default cluster, overridden by the WASM filter
```
Style: `WASM filter` -> `Proxy-Wasm plugin`

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,73 @@
+## Proxy-Wasm plugin example: Envoy TCP Routing
+
+Proxy-Wasm TCP filter that dynamically routes connections to different upstream clusters based on the source IP address.
```
Style: `filter` -> `plugin` (here and throughout the PR when referring to compiled plugin and not Envoy filter itself)

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,73 @@
+
+### Building
+
+Build the WASM plugin from the example directory:
```
Nit: We don't have this line in other examples, since they are considered standalone (and `README.md` is already in the example's directory, so it feel redundant). While I'm not strongly against having such line in examples, I value consistent style across them.

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,73 @@
+
+This example can be run with [`docker compose`](https://docs.docker.com/compose/install/) and has a matching Envoy configuration.
+
+From the example directory:
```
Nit: We don't have this line in other examples, since they are considered standalone (and `README.md` is already in the example's directory, so it feel redundant). While I'm not strongly against having such line in examples, I value consistent style across them.

PiotrSikora commented on examples/envoy_tcp_routing/docker-compose.yaml:
```diff
@@ -0,0 +1,47 @@
+    image: envoyproxy/envoy:v1.34-latest
+    hostname: envoy
+    command:
+      - /usr/local/bin/envoy
```
The point of `command` is that you don't need to hardcode the binary path, so this is unnecessary.

Also, you might want to combine some args and use the full argument name for readability:
```
command:
  - --config-path /etc/envoy/envoy.yaml
  - --log-level info
  - --service-cluster proxy
```

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
```diff
@@ -0,0 +1,73 @@
+
+### Note on Destination Information
+
+In this example, both Envoy clusters (`egress-router1` and `egress-router2`) target the local `httpbin` service. In deployments where original destination metadata must be preserved across TCP hops, the [PROXY protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) is one practical option.
```
This feels misplaced at the end. Perhaps move it to the intro or overview and preface it with "In production deployments..."?

PiotrSikora replied:
I ended up dropping this altogether, since the upstream needs to be configured to accepts PROXY protocol from the proxy, so we'd need to get into nitty gritty details here.

SiiiTschiii replied:
@PiotrSikora  Agreed on dropping that note since context is lacking. To still show a production-grade use case without the configuration baggage, we could add a 'transparent proxy' example using PROXY Protocol, Kind and Istio.

We could polish and adapt this example: https://github.com/SiiiTschiii/wasmerang/tree/main/examples/istio-k8s

Needless to say: Not in this PR ;-)

PiotrSikora replied:
> We could polish and adapt this example: https://github.com/SiiiTschiii/wasmerang/tree/main/examples/istio-k8s
> 
> Needless to say: Not in this PR ;-)

We could also just link to `wasmerang` repo from the `README.md`, maybe in a new section called `Real world plugins` or `Production plugins` or something like that.

PiotrSikora commented on examples/envoy_tcp_routing/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+name = "proxy-wasm-example-envoy-tcp-routing"
+version = "0.0.1"
+authors = ["Proxy-Wasm contributors"]
+description = "Proxy-Wasm plugin example: Envoy TCP Routing based on source IP"
```
Nit: `Proxy-Wasm plugin example: Envoy TCP routing`

PiotrSikora commented on examples/envoy_tcp_routing/envoy.yaml:
```diff
@@ -0,0 +1,83 @@
+                      "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
+                      path: /dev/stdout
+                      log_format:
+                        text_format: "[%START_TIME%] cluster=%UPSTREAM_CLUSTER% src=%DOWNSTREAM_REMOTE_ADDRESS% dst=%DOWNSTREAM_LOCAL_ADDRESS% -> %UPSTREAM_HOST%\n"
```
1. `dst` usually refers to the destination, not local listening address, so I replaced it with `via`.
2. moved `cluster` next to the target destination, and dropped `cluster` to make it proxy agnostic.

```suggestion
                        text_format: "[%START_TIME%] src=%DOWNSTREAM_REMOTE_ADDRESS% via=%DOWNSTREAM_LOCAL_ADDRESS% dst=%UPSTREAM_HOST% (%UPSTREAM_CLUSTER%)\n"
```

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
````diff
@@ -0,0 +1,73 @@
+
+In a separate terminal test the routing behavior with different source IP addresses:
+
+```bash
````
Nit: `bash` -> `sh`

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
````diff
@@ -0,0 +1,73 @@
+
+**For even IP (last octet 10) → routes via egress-router1:**
+
+```
````
Missing `console` style annotation.

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
````diff
@@ -0,0 +1,73 @@
+
+**For odd IP (last octet 11) → routes via egress-router2:**
+
+```
````
Missing `console` style annotation.

PiotrSikora commented on examples/envoy_tcp_routing/README.md:
````diff
@@ -0,0 +1,73 @@
+$ docker compose up
+```
+
+### Test the Routing
````
Could you convert this into 2 groups like in other examples (e.g. see `http_auth_random/README.md`)?
```
#### Routing even source IPs

Send HTTP request to ...:
```sh
docker run ...
```

Expected logs:
```console
[...] TCP WASM] Routed source 172.22.0.10:39484 to egress-router1
[...] [2026-05-13T03:08:18.423Z] cluster=egress-router1 src=172.22.0.10:39484 dst=172.22.0.2:10000 -> 172.22.0.3:80
```

While I don't have particular preference for one style vs other, there is no reason to make everything different than in existing examples.

PiotrSikora reviewed (commented):
@SiiiTschiii that should be the last round of it, pretty much only style fixes.

PiotrSikora committed:
review: style.

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora committed:
Merge remote-tracking branch 'origin/main' into SiiiTschiii/example-tcp-re-routing

Signed-off-by: Piotr Sikora <code@piotrsikora.dev>

PiotrSikora reviewed (approved):
Thank you!
